ArduCopter/Copter.cpp

```diff
@@ -773,6 +779,31 @@ bool Copter::get_rate_ef_targets(Vector3f& rate_ef_targets) const
     return true;
 }
 
+void Copter::check_outdoors_ready() {
+    if (!copter.ahrs.get_option_90_active())
+    {
+        // if (!(arming.armed)) { //Only update mode when disarmed
+        bool optflow_good = optflow.enabled() && optflow.healthy();
+        bool gps_good = arming.gps_checks_indoor_mode(false) && (gps.num_sats() >= 9 && gps.get_hdop() < 140);
+        GCS_SEND_TEXT(MAV_SEVERITY_INFO, "%d %d", gps_good, optflow_good);
+        bool new_outdoors_ready = gps_good || !optflow_good;
+        if (outdoors_ready != new_outdoors_ready) {
+            outdoors_ready = new_outdoors_ready;
+            if (new_outdoors_ready) {
+                ahrs.set_posvelyaw_source_set(1);
```

Comment:
2. If outdoors ready has changed then we update the ahrs source and set the outdoors_ready variable.